[quant] Fix conv packed param serialization in state_dict

jerryzh168 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #52802 [do not land] ad hoc test for conv serialization
* #51639 [quant][graphmode][fx] Add support for packed params in state_dict
* **#52787 [quant] Fix conv packed param serialization in state_dict**

Summary:
Current conv packed param can be serialized/deserialized with `torch.jit.save/torch.jit.load`, but
can't be saved with `torch.load(m.state_dict())/torch.save(m.state_dict())`

reason is (from James):
```
I think the issue probably has to do with the normal pickle deserialization not detecting List[Optional[Tensor]] if it doesn't witness a None in the list. IIRC this is implemented on the TorchScript side through this type tag mechanism: https://github.com/.../jit/serialization/unpickler.cpp...
```

This PR is a hack but acceptable to JIT team until a proper solution is proposed.

Test Plan:
Will be tested in next PR

BC test (from: https://github.com/pytorch/pytorch/commit/b3f0297a94977636fd90c0fe6fa9b971ff9f81e2):
```
python test/test_quantization.py TestSerialization.test_conv2d_graph_v2
python test/test_quantization.py TestSerialization.test_conv2d_nobias_graph_v2
```

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D26649272](https://our.internmc.facebook.com/intern/diff/D26649272)

jerryzh168 committed:
[quant] Fix conv packed param serialization in state_dict

Summary:
Current conv packed param can be serialized/deserialized with `torch.jit.save/torch.jit.load`, but
can't be saved with `torch.load(m.state_dict())/torch.save(m.state_dict())`

reason is (from James):
```
I think the issue probably has to do with the normal pickle deserialization not detecting List[Optional[Tensor]] if it doesn't witness a None in the list. IIRC this is implemented on the TorchScript side through this type tag mechanism: https://github.com/.../jit/serialization/unpickler.cpp...
```

This PR is a hack but acceptable to JIT team until a proper solution is proposed.

Test Plan:
Will be tested in next PR

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

vkuzo reviewed (approved):
lg if CI is green, thanks!

IMO this is a medium to high risk PR.  Would it be possible to attach a paste of why the "before" code didn't work to the PR summary, and could we modify the test plan to show that we can run the conv BC tests successfully, with the actual test commands?  Motivation: making it easier for an oncall from another team to see if this PR caused any problems.

jerryzh168 commented:
> lg if CI is green, thanks!
> 
> IMO this is a medium to high risk PR. Would it be possible to attach a paste of why the "before" code didn't work to the PR summary, and could we modify the test plan to show that we can run the conv BC tests successfully, with the actual test commands? Motivation: making it easier for an oncall from another team to see if this PR caused any problems.

re: why the "before" code didn't work, I have quoted from James' comment, is that good enough?

jerryzh168 committed:
Update on "[quant] Fix conv packed param serialization in state_dict"


Summary:
Current conv packed param can be serialized/deserialized with `torch.jit.save/torch.jit.load`, but
can't be saved with `torch.load(m.state_dict())/torch.save(m.state_dict())`

reason is (from James):
```
I think the issue probably has to do with the normal pickle deserialization not detecting List[Optional[Tensor]] if it doesn't witness a None in the list. IIRC this is implemented on the TorchScript side through this type tag mechanism: https://github.com/.../jit/serialization/unpickler.cpp...
```

This PR is a hack but acceptable to JIT team until a proper solution is proposed.

Test Plan:
Will be tested in next PR

BC test (from: https://github.com/pytorch/pytorch/commit/b3f0297a94977636fd90c0fe6fa9b971ff9f81e2):
```
python test/test_quantization.py TestSerialization.test_conv2d_graph_v2
python test/test_quantization.py TestSerialization.test_conv2d_nobias_graph_v2
```

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D26649272](https://our.internmc.facebook.com/intern/diff/D26649272)

[ghstack-poisoned]
